File: compilers/java/src/test/java/runar/compiler/passes/EmbedAlwaysTest.java
package runar.compiler.passes;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.Test;
import runar.compiler.frontend.ParserDispatch;
import runar.compiler.ir.ast.ContractNode;
import runar.compiler.ir.ast.PropertyNode;

/**
 * Issue #109 — {@code /** @embedAlways *&#47;} readonly-field DCE opt-out. Port
 * of embed-always.test.ts: parser flag, DCE preservation into a constructor
 * slot (byte parity with the TypeScript reference), and the strip warning.
 */
class EmbedAlwaysTest {

    private static String source(String directive) {
        return "class Meta extends SmartContract {\n"
            + "  readonly pubKeyHash: Addr;\n"
            + "  " + directive + "\n"
            + "  readonly metadataId: ByteString;\n"
            + "  constructor(pubKeyHash: Addr, metadataId: ByteString) {\n"
            + "    super(pubKeyHash, metadataId); this.pubKeyHash = pubKeyHash; this.metadataId = metadataId;\n"
            + "  }\n"
            + "  public unlock(sig: Sig, pubKey: PubKey) {\n"
            + "    assert(hash160(pubKey) === this.pubKeyHash); assert(checkSig(sig, pubKey));\n"
            + "  }\n"
            + "}\n";
    }

    private static PropertyNode prop(ContractNode c, String name) {
        return c.properties().stream().filter(p -> p.name().equals(name)).findFirst().orElseThrow();
    }

    // ---- Parser ---------------------------------------------------------------

    @Test
    void setsEmbedAlwaysOnJsdocDirective() throws Exception {
        ContractNode c = ParserDispatch.parse(source("/** @embedAlways */"), "Meta.runar.ts");
        assertTrue(prop(c, "metadataId").embedAlways());
        assertFalse(prop(c, "pubKeyHash").embedAlways());
    }

    @Test
    void recognizesLineCommentDirective() throws Exception {
        ContractNode c = ParserDispatch.parse(source("// @embedAlways"), "Meta.runar.ts");
        assertTrue(prop(c, "metadataId").embedAlways());
    }

    @Test
    void leavesEmbedAlwaysUnsetWithNoDirective() throws Exception {
        ContractNode c = ParserDispatch.parse(source(""), "Meta.runar.ts");
        assertFalse(prop(c, "metadataId").embedAlways());
    }

    // ---- Preservation (byte parity with the TypeScript reference) ------------

    // Fold-OFF locking-script templates captured from the TS compiler
    // (packages/runar-compiler) compile(src, { disableConstantFolding: true }).scriptHex.
    private static final String TS_PLAIN_HEX = "76a90088ac";
    private static final String TS_EMBED_HEX = "0078a900887b7bac77";

    @Test
    void unannotatedFieldIsEliminated() throws Exception {
        String plain = PipelineTestSupport.hex(source(""), "Meta.runar.ts");
        assertEquals(TS_PLAIN_HEX, plain, "un-annotated template must match the TS reference");
    }

    @Test
    void annotatedFieldIsPreservedByteIdenticalToTs() throws Exception {
        String embed = PipelineTestSupport.hex(source("/** @embedAlways */"), "Meta.runar.ts");
        assertEquals(TS_EMBED_HEX, embed, "@embedAlways template must match the TS reference");
    }

    @Test
    void annotatedHexCarriesMoreBytesThanUnannotated() throws Exception {
        String plain = PipelineTestSupport.hex(source(""), "Meta.runar.ts");
        String embed = PipelineTestSupport.hex(source("/** @embedAlways */"), "Meta.runar.ts");
        assertNotEquals(plain, embed);
        assertTrue(embed.length() > plain.length(), "embed template must carry the extra field bytes");
    }

    // ---- Warning (Option 4) ---------------------------------------------------

    @Test
    void warnsForEliminatedUnannotatedReadonlyField() throws Exception {
        var diags = PipelineTestSupport.diagnostics(source(""), "Meta.runar.ts");
        assertTrue(diags.warnings().stream().anyMatch(w ->
            w.contains("metadataId") && w.contains("eliminated by DCE") && w.contains("@embedAlways")),
            diags.warnings().toString());
    }

    @Test
    void doesNotWarnWhenFieldIsAnnotated() throws Exception {
        var diags = PipelineTestSupport.diagnostics(source("/** @embedAlways */"), "Meta.runar.ts");
        assertFalse(diags.warnings().stream().anyMatch(w -> w.contains("metadataId")),
            diags.warnings().toString());
    }

    @Test
    void doesNotWarnForReferencedReadonlyField() throws Exception {
        String referenced =
            "class P2PKH extends SmartContract {\n"
            + "  readonly pubKeyHash: Addr;\n"
            + "  constructor(pubKeyHash: Addr) { super(pubKeyHash); this.pubKeyHash = pubKeyHash; }\n"
            + "  public unlock(sig: Sig, pubKey: PubKey) {\n"
            + "    assert(hash160(pubKey) === this.pubKeyHash); assert(checkSig(sig, pubKey));\n"
            + "  }\n"
            + "}\n";
        var diags = PipelineTestSupport.diagnostics(referenced, "P2PKH.runar.ts");
        assertFalse(diags.warnings().stream().anyMatch(w -> w.contains("pubKeyHash")),
            diags.warnings().toString());
    }
}

File: compilers/go/compiler/embed_always_test.go
package compiler

import (
	"strings"
	"testing"
)

// Issue #109 — `/** @embedAlways */` readonly-field DCE opt-out (Go tier).
//
// A readonly field no method references is normally eliminated by ANF
// dead-binding DCE (the dead load_prop is dropped, so no constructor slot is
// emitted), silently removing deploy-time metadata an author intends to recover
// from the on-chain script later. The @embedAlways directive forces the field
// into the locking script (a constructor slot). Golden hexes are the TypeScript
// reference output (fold-OFF) — the Go tier must be byte-identical.

func metaSource(directive string) string {
	return `
import { SmartContract, assert, Addr, PubKey, Sig, ByteString, hash160, checkSig } from 'runar-lang';
class Meta extends SmartContract {
  readonly pubKeyHash: Addr;
  ` + directive + `
  readonly metadataId: ByteString;
  constructor(pubKeyHash: Addr, metadataId: ByteString) {
    super(pubKeyHash, metadataId);
    this.pubKeyHash = pubKeyHash;
    this.metadataId = metadataId;
  }
  public unlock(sig: Sig, pubKey: PubKey) {
    assert(hash160(pubKey) === this.pubKeyHash);
    assert(checkSig(sig, pubKey));
  }
}
`
}

// TS-reference golden scripts (fold-OFF).
const (
	metaPlainGoldenScript = "76a90088ac"
	metaEmbedGoldenScript = "0078a900887b7bac77"
)

func TestEmbedAlways_UnannotatedFieldEliminated(t *testing.T) {
	r := CompileFromSourceStrWithResult(metaSource(""), "Meta.runar.ts", CompileOptions{DisableConstantFolding: true})
	if !r.Success {
		t.Fatalf("compile failed: %v", diagMessages(r))
	}
	if r.ScriptHex != metaPlainGoldenScript {
		t.Fatalf("plain script mismatch:\n  got %s\n  want %s", r.ScriptHex, metaPlainGoldenScript)
	}
	names := slotNames(r)
	if contains(names, "metadataId") {
		t.Fatalf("un-annotated metadataId should be eliminated; slots=%v", names)
	}
	if !contains(names, "pubKeyHash") {
		t.Fatalf("pubKeyHash slot missing; slots=%v", names)
	}
}

func TestEmbedAlways_AnnotatedFieldPreservedByteIdenticalToTS(t *testing.T) {
	r := CompileFromSourceStrWithResult(metaSource("/** @embedAlways */"), "Meta.runar.ts", CompileOptions{DisableConstantFolding: true})
	if !r.Success {
		t.Fatalf("compile failed: %v", diagMessages(r))
	}
	if r.ScriptHex != metaEmbedGoldenScript {
		t.Fatalf("embed script mismatch (Go != TS):\n  got %s\n  want %s", r.ScriptHex, metaEmbedGoldenScript)
	}
	names := slotNames(r)
	// Both slots present; metadataId first (injected preservation load precedes
	// the pubKeyHash read in unlock) — matches TS constructorSlots order.
	if len(names) != 2 || names[0] != "metadataId" || names[1] != "pubKeyHash" {
		t.Fatalf("constructor slot order mismatch; got %v want [metadataId pubKeyHash]", names)
	}
}

func TestEmbedAlways_LineCommentForm(t *testing.T) {
	r := CompileFromSourceStrWithResult(metaSource("// @embedAlways"), "Meta.runar.ts", CompileOptions{DisableConstantFolding: true})
	if !r.Success {
		t.Fatalf("compile failed: %v", diagMessages(r))
	}
	if r.ScriptHex != metaEmbedGoldenScript {
		t.Fatalf("line-comment embed script mismatch:\n  got %s\n  want %s", r.ScriptHex, metaEmbedGoldenScript)
	}
}

func TestEmbedAlways_WarnsOnStrippedUnannotatedField(t *testing.T) {
	r := CompileFromSourceStrWithResult(metaSource(""), "Meta.runar.ts", CompileOptions{DisableConstantFolding: true})
	if !r.Success {
		t.Fatalf("compile failed: %v", diagMessages(r))
	}
	if !anyWarnContains(r, "metadataId") || !anyWarnContains(r, "@embedAlways") {
		t.Fatalf("expected DCE warning naming metadataId + @embedAlways; diags=%v", diagMessages(r))
	}
}

func TestEmbedAlways_NoWarnWhenAnnotated(t *testing.T) {
	r := CompileFromSourceStrWithResult(metaSource("/** @embedAlways */"), "Meta.runar.ts", CompileOptions{DisableConstantFolding: true})
	if anyWarnContains(r, "metadataId") {
		t.Fatalf("annotated field should not warn; diags=%v", diagMessages(r))
	}
}

// --- helpers ---

func slotNames(r *CompileResult) []string {
	if r.Artifact == nil {
		return nil
	}
	params := r.Artifact.ABI.Constructor.Params
	out := make([]string, 0, len(r.Artifact.ConstructorSlots))
	for _, s := range r.Artifact.ConstructorSlots {
		if s.ParamIndex >= 0 && s.ParamIndex < len(params) {
			out = append(out, params[s.ParamIndex].Name)
		}
	}
	return out
}

func contains(xs []string, x string) bool {
	for _, s := range xs {
		if s == x {
			return true
		}
	}
	return false
}

func diagMessages(r *CompileResult) []string {
	out := make([]string, 0, len(r.Diagnostics))
	for _, d := range r.Diagnostics {
		out = append(out, string(d.Severity)+": "+d.Message)
	}
	return out
}

func anyWarnContains(r *CompileResult, sub string) bool {
	for _, d := range r.Diagnostics {
		if d.Severity == "warning" && strings.Contains(d.Message, sub) {
			return true
		}
	}
	return false
}
